File: iOS/Sources/VertuEdge/RPA/VertuFlowEngine.swift
import Foundation
import UIKit

actor VertuFlowEngine {
    private let parser = FlowParser()

    func execute(yaml: String) async -> FlowResult {
        do {
            let flow = try parser.parse(yaml)
            var logs: [String] = ["Starting flow: \(flow.name ?? flow.appId)"]
            for action in flow.actions {
                let log = await executeAction(action)
                logs.append(log)
            }
            return .success(logs: logs)
        } catch {
            return .error(message: "Parse error: \(error.localizedDescription)")
        }
    }

    private func executeAction(_ action: FlowAction) async -> String {
        switch action {
        case .launchApp(let appId):
            await MainActor.run {
                if let url = URL(string: appId) {
                    UIApplication.shared.open(url)
                }
            }
            return "Launch app: \(appId)"
        case .tapOn(let selector, _):
            return "Tap on: \(selector)"
        case .inputText(let text):
            return "Input text: \(text)"
        case .assertVisible(let selector):
            return "Assert visible: \(selector)"
        case .assertNotVisible(let selector):
            return "Assert not visible: \(selector)"
        case .scrollUntilVisible(let selector, _):
            return "Scroll until visible: \(selector)"
        case .openLink(let url):
            await MainActor.run {
                if let u = URL(string: url) {
                    UIApplication.shared.open(u)
                }
            }
            return "Open link: \(url)"
        case .wait(let ms):
            try? await Task.sleep(nanoseconds: UInt64(ms) * 1_000_000)
            return "Wait: \(ms)ms"
        case .pressKey(let key):
            return "Press key: \(key)"
        case .runAiPrompt(let prompt, _):
            return "Run AI prompt: \(prompt)"
        case .webAction(let action, let selector, _):
            return "Web action: \(action) on \(selector ?? "page")"
        case .takeScreenshot(let label):
            return "Screenshot: \(label ?? "unnamed")"
        case .clearState(let appId):
            return "Clear state: \(appId)"
        }
    }
}

enum FlowResult {
    case success(logs: [String])
    case error(message: String)
}

File: Android/app/src/main/java/com/vertu/edge/rpa/VertuFlowEngine.kt
package com.vertu.edge.rpa

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import kotlinx.coroutines.delay
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class VertuFlowEngine @Inject constructor(
    private val context: Context
) {
    private val tag = "VertuFlowEngine"
    private val parser = FlowParser()

    suspend fun execute(yaml: String): FlowResult {
        val parseResult = parser.parse(yaml)
        if (parseResult.isFailure) {
            return FlowResult.Error("Parse error: ${parseResult.exceptionOrNull()?.message}")
        }
        val flow = parseResult.getOrThrow()
        val logs = mutableListOf<String>()
        logs.add("Starting flow: ${flow.name ?: flow.appId}")

        for (action in flow.actions) {
            val result = executeAction(action, logs)
            if (result is ActionResult.Error) {
                return FlowResult.Error("Action failed: ${result.message}")
            }
        }
        return FlowResult.Success(logs)
    }

    private suspend fun executeAction(action: FlowAction, logs: MutableList<String>): ActionResult {
        return when (action) {
            is FlowAction.LaunchApp -> {
                logs.add("Launching app: ${action.appId}")
                launchApp(action.appId)
            }
            is FlowAction.TapOn -> {
                logs.add("Tap on: ${action.selector}")
                ActionResult.Success
            }
            is FlowAction.InputText -> {
                logs.add("Input text: ${action.text}")
                ActionResult.Success
            }
            is FlowAction.AssertVisible -> {
                logs.add("Assert visible: ${action.selector}")
                ActionResult.Success
            }
            is FlowAction.AssertNotVisible -> {
                logs.add("Assert not visible: ${action.selector}")
                ActionResult.Success
            }
            is FlowAction.ScrollUntilVisible -> {
                logs.add("Scroll until visible: ${action.selector}")
                ActionResult.Success
            }
            is FlowAction.OpenLink -> {
                logs.add("Open link: ${action.url}")
                openLink(action.url)
            }
            is FlowAction.Wait -> {
                logs.add("Wait: ${action.durationMs}ms")
                delay(action.durationMs)
                ActionResult.Success
            }
            is FlowAction.PressKey -> {
                logs.add("Press key: ${action.key}")
                ActionResult.Success
            }
            is FlowAction.RunAiPrompt -> {
                logs.add("Run AI prompt: ${action.prompt}")
                ActionResult.Success
            }
            is FlowAction.WebAction -> {
                logs.add("Web action: ${action.action} on ${action.selector}")
                ActionResult.Success
            }
            is FlowAction.TakeScreenshot -> {
                logs.add("Take screenshot: ${action.label ?: "unnamed"}")
                ActionResult.Success
            }
            is FlowAction.ClearState -> {
                logs.add("Clear state: ${action.appId}")
                ActionResult.Success
            }
        }
    }

    private fun launchApp(appId: String): ActionResult {
        return try {
            if (appId.isNotEmpty()) {
                val intent = context.packageManager.getLaunchIntentForPackage(appId)
                    ?: return ActionResult.Error("App not found: $appId")
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(intent)
            }
            ActionResult.Success
        } catch (e: Exception) {
            Log.e(tag, "Failed to launch app: $appId", e)
            ActionResult.Error(e.message ?: "Unknown error launching app")
        }
    }

    private fun openLink(url: String): ActionResult {
        return try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
            ActionResult.Success
        } catch (e: Exception) {
            ActionResult.Error(e.message ?: "Unknown error opening link")
        }
    }
}

sealed class ActionResult {
    data object Success : ActionResult()
    data class Error(val message: String) : ActionResult()
}

sealed class FlowResult {
    data class Success(val logs: List<String>) : FlowResult()
    data class Error(val message: String) : FlowResult()
}
